build: Getting ready to Qt 6 (4/n). Improve `build-aux/m4/bitcoin_qt.m4`

hebasto opened the pull request:
This PR adds a test macro `_BITCOIN_QT_CHECK_APP`, which ensures that Qt resource system object (`*.rcc`) have being compiled and linked properly. Found it such checks very useful while working on bitcoin/bitcoin#24798.

Here are examples from `configure` logs:
- successful check:
```
...
checking for Qt5Core >= 5.11.3... yes
checking for QCoreApplication initialization... yes
checking for Qt5Gui >= 5.11.3... yes
checking for QGuiApplication initialization... yes
checking for Qt5Widgets >= 5.11.3... yes
checking for QApplication initialization... yes
checking for Qt5Network >= 5.11.3... yes
checking for Qt5Test >= 5.11.3... yes
checking for Qt5DBus >= 5.11.3... yes
...
```

- failed check:
```
...
checking for QCoreApplication initialization... no
configure: WARNING: QCoreApplication failed to initialize.; bitcoin-qt frontend will not be built
checking whether to build Bitcoin Core GUI... no
...
```

The fourth commit may fix build in some weird setups, but it is definitely required for Qt 6.

hebasto committed:
build: Check Qt for PIE/PIC earlier

This is a prerequisite for the following commit.

A move-only commit, use `diff --color-moved=dimmed-zebra` to verify.

hebasto committed:
build, qt: Add `_BITCOIN_QT_CHECK_APP` macro

This change increases robustness of the build system. A passed test
guarantees that the required Qt resource system object have being linked
properly.

hebasto committed:
build, qt, doc: Improve `_BITCOIN_QT_CHECK_STATIC_PLUGIN` description

Due to appending `qt_lib_suffix`, only one library can be passed to
the `_BITCOIN_QT_CHECK_STATIC_PLUGIN` macro.

hebasto committed:
build: Pass `qt_lib_path` from depends to the `configure` script

This change makes behavior consistent among `with_qt_*` variables, and
it is required for Qt 6.

dunxen reviewed (commented):
tACK b3a6ca2

Confirmed that the new checks are successfully run with configure.

hebasto commented:
Updated b3a6ca23d3e55a81bee573b92e3946a284f05761 -> 43af01e23b5e7e5afba2da2863aafade6f8ad752 ([pr24813.02](https://github.com/hebasto/bitcoin/commits/pr24813.02) -> [pr24813.03](https://github.com/hebasto/bitcoin/commits/pr24813.03), [diff](https://github.com/hebasto/bitcoin/compare/pr24813.02..pr24813.03)):

- addressed @dunxen's [comment](https://github.com/bitcoin/bitcoin/pull/24813#discussion_r849263257)

hebasto commented on build-aux/m4/bitcoin_qt.m4:
```diff
@@ -312,12 +313,12 @@ AC_DEFUN([_BITCOIN_QT_IS_STATIC],[
 dnl Requires: INCLUDES and LIBS must be populated as necessary.
 dnl Inputs: $1: A static plugin name.
-dnl Inputs: $2: The libraries that resolve $1.
+dnl Inputs: $2: The library that resolve $1.
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/24813#issuecomment-1097888420).

dunxen commented:
reACK 43af01e

fanquake commented:
> which ensures that Qt resource system object (*.rcc) have being compiled and linked properly. Found it such checks very useful while working on https://github.com/bitcoin/bitcoin/pull/24798.

Can you elaborate on when a user or developer would actually run into the scenario of these resources not being linked properly? I'm not against adding additional tests that might be useful to others, but if these are checks that were only helpful in your dev environment while things were broken / you were debugging, then it's less clear they should be added to configure.

hebasto commented:
> > which ensures that Qt resource system object (*.rcc) have being compiled and linked properly. Found it such checks very useful while working on #24798.
> 
> Can you elaborate on when a user or developer would actually run into the scenario of these resources not being linked properly? I'm not against adding additional tests that might be useful to others, but if these are checks that were only helpful in your dev environment while things were broken / you were debugging, then it's less clear they should be added to configure.

For example, in 9f61d6a7605eee4b116b8791276bd71baf75bf10 forget to include `${qt_lib_path}/objects-Release/Widgets_resources_1/.rcc/qrc_qstyle.cpp.o`

fanquake commented:
> For example, in https://github.com/bitcoin/bitcoin/commit/9f61d6a7605eee4b116b8791276bd71baf75bf10 forget to include ${qt_lib_path}/objects-Release/Widgets_resources_1/.rcc/qrc_qstyle.cpp.o

In that case, concept ~0 for the new checks. They seem to be a symptom of other hacks (trying to link against specific .o files directly in configure checks? I'll comment in #24798), which I would hope have a better solution, and this isn't really guarding against a situation that someone is going to run into in the wild.

hebasto commented:
> ... which I would hope have a better solution

Me too :)

Not talking about a "better" solution, but here are some alternative ones:
- wait for Qt6 patch for `*.pc` files which works with static builds
- develop our own Qt6 patch for `*.pc` files which works with static builds
- use [`CMake`](https://github.com/bitcoin/bitcoin/issues/24911) for Qt6 builds

fanquake commented:
> Not talking about a "better" solution, but here are some alternative ones:

That discussion might be better saved for #24798. I'm a Concept NACK here. I don't think these tests should be added, as they are guarding against situations that, from what I understand, can only could occur if we introduce the [fairly fragile workarounds](https://github.com/bitcoin/bitcoin/pull/24813#issuecomment-1103791264) from #24798 into our code (it would have been handy if this was outlined in the PR description), which I don't think we should do.

fanquake commented:
I think this should be closed / combined into #24798 for now. In any case, I don't think this is the approach (linking/testing object files) we would want to use in our build system to add Qt 6 support.

hebasto commented:
> I think this should be closed / combined into #24798 for now.

Agree.

> In any case, I don't think this is the approach (linking/testing object files) we would want to use in our build system to add Qt 6 support.

Agree again. Moving to CMake-based build system [fixes](https://github.com/bitcoin/bitcoin/pull/25797) it :tiger2:
